src/validate.rs

```diff
@@ -1,23 +1,723 @@
+            owner.make_ascii_lowercase();
+            context.update(&owner[..self.owner.len()]);
+        } else {
+            context.update(self.owner.as_slice());
+        }
+
+        // Add basic DNSKEY fields.
+        context.update(&self.flags.to_be_bytes());
+        context.update(&[3, self.algorithm().to_int()]);
+
+        // Add the public key.
+        self.key.digest(&mut context);
+
+        // Finalize the digest.
+        let digest = context.finish().as_ref().into();
+        Ok(Ds::new(self.key_tag(), self.algorithm(), algorithm, digest)
+            .unwrap())
+    }
+}
+
+//--- Conversion to and from DNSKEYs
+
+impl<Octs: AsRef<[u8]>> Key<Octs> {
+    /// Deserialize a key from DNSKEY record data.
+    ///
+    /// # Errors
+    ///
+    /// Fails if the DNSKEY uses an unknown protocol or contains an invalid
+    /// public key (e.g. one of the wrong size for the signature algorithm).
+    pub fn from_dnskey(
+        owner: Name<Octs>,
+        dnskey: Dnskey<Octs>,
+    ) -> Result<Self, FromDnskeyError> {
+        if dnskey.protocol() != 3 {
+            return Err(FromDnskeyError::UnsupportedProtocol);
+        }
+
+        let flags = dnskey.flags();
+        let algorithm = dnskey.algorithm();
+        let key = dnskey.public_key().as_ref();
+        let key = RawPublicKey::from_dnskey_format(algorithm, key)?;
+        Ok(Self { owner, flags, key })
+    }
+
+    /// Parse a DNSSEC key from a DNSKEY record in presentation format.
+    ///
+    /// This format is popularized for storing alongside private keys by the
+    /// BIND name server.  This function is convenient for loading such keys.
+    ///
+    /// The text should consist of a single line of the following format (each
+    /// field is separated by a non-zero number of ASCII spaces):
+    ///
+    /// ```text
+    /// <domain-name> <record-class> DNSKEY <record-data> [<comment>]
+    /// ```
+    ///
+    /// Where `<record-data>` consists of the following fields:
+    ///
+    /// ```text
+    /// <flags> <protocol> <algorithm> <encoded-public-key>
+    /// ```
+    ///
+    /// The first three fields are simple integers, while the last field is
+    /// Base64 encoded data (with or without padding).  The [`from_dnskey()`]
+    /// and [`to_dnskey()`] read from and serialize to the Base64-decoded data
+    /// format.
+    ///
+    /// [`from_dnskey()`]: Self::from_dnskey()
+    /// [`to_dnskey()`]: Self::to_dnskey()
+    ///
+    /// The `<comment>` is any text starting with an ASCII semicolon.
+    pub fn parse_dnskey_text(
+        dnskey: &str,
+    ) -> Result<Self, ParseDnskeyTextError>
+    where
+        Octs: FromBuilder,
+        Octs::Builder: EmptyBuilder + Composer,
+    {
+        // Ensure there is a single line in the input.
+        let (line, rest) = dnskey.split_once('\n').unwrap_or((dnskey, ""));
+        if !rest.trim().is_empty() {
+            return Err(ParseDnskeyTextError::Misformatted);
+        }
```

Comment:
This code failed to parse files produced by BIND `dnssec-keygen` which have leading comment lines, e.g.:

```
$ cat /tmp/x/Kexample.com.+013+04622.key 
; This is a key-signing key, keyid 4622, for example.com.
; Created: 20241015203641 (Tue Oct 15 22:36:41 2024)
; Publish: 20241015203641 (Tue Oct 15 22:36:41 2024)
; Activate: 20241015203641 (Tue Oct 15 22:36:41 2024)
example.com. IN DNSKEY 257 3 13 e9Usj72J/s5jY9Rxi+kc9M7n3f1osxzVjD3HaS9GYTHqjDNolHraAevV RjH8pj+I7JtcT/pE1oqwxnxC0pKeAw==
```

The following modified code worked for me:

```suggestion
        // Skip leading comment lines (BIND uses these to record key timing
        // data).
        let mut line = dnskey;
        while let Some((this_line, rest)) = line.split_once('\n') {
            if !this_line.starts_with(';') {
                // Ensure there is a single data line in the input.
                if !rest.trim().is_empty() {
                    return Err(ParseDnskeyTextError::Misformatted);
                } else {
                    line = this_line;
                    break;
                }
            } else {
                line = rest;
            }
        }
```

Reply:
Thanks!  I'll make it a bit more resilient too (to skip whitespace preceding the colon).